engine/jobmaster/dm/metadata/ddl.go

```diff
@@ -14,35 +14,195 @@
 
-// DDLStore manages the state of ddls.
-// Write by DDLCoordinator.
-type DDLStore struct {
+// TargetTable represents a downstream table.
+type TargetTable struct {
+	Schema string
+	Table  string
+}
+
+// MarshalText implements json.MarshalText.
+func (tt TargetTable) MarshalText() (text []byte, err error) {
+	type t TargetTable
+	return json.Marshal(t(tt))
+}
+
+// UnmarshalText implements json.UnmarshalText.
+func (tt *TargetTable) UnmarshalText(text []byte) error {
+	type t TargetTable
+	return json.Unmarshal(text, (*t)(tt))
+}
+
+// DDLType defines the type of DDL
+type DDLType uint64
+
+// Defines DDLType here.
+const (
+	CreateTable DDLType = iota + 1
+	DropTable
+	OtherDDL
+)
+
+// DDLItem represents a DDL item
+type DDLItem struct {
+	TargetTable TargetTable
+	SourceTable SourceTable
+	Tables      []string
+	DDLs        []string
+	Type        DDLType
+}
+
+// DroppedColumns represents the state of dropped columns
+type DroppedColumns struct {
+	// column -> source table
+	Cols map[string]map[SourceTable]struct{}
+}
+
+// DroppedColumnsStore manages the dropped columns state
+type DroppedColumnsStore struct {
 	*frameworkMetaStore
+
+	mu          sync.Mutex
+	targetTable TargetTable
 }
 
-// NewDDLStore returns a new DDLStore instance
-func NewDDLStore(kvClient metaModel.KVClient) *DDLStore {
-	ddlStore := &DDLStore{
-		frameworkMetaStore: newTOMLFrameworkMetaStore(kvClient),
+// NewDroppedColumnsStore returns a new DroppedColumnsStore instance
+func NewDroppedColumnsStore(kvClient metaModel.KVClient, targetTable TargetTable) *DroppedColumnsStore {
+	droppedColumnStore := &DroppedColumnsStore{
+		frameworkMetaStore: newJSONFrameworkMetaStore(kvClient),
+		targetTable:        targetTable,
 	}
-	ddlStore.frameworkMetaStore.stateFactory = ddlStore
-	return ddlStore
+	droppedColumnStore.frameworkMetaStore.stateFactory = droppedColumnStore
+	return droppedColumnStore
 }
 
-// CreateState creates an empty DDL object
-func (ddlStore *DDLStore) createState() state {
-	return &DDL{}
+// CreateState creates an empty DroppedColumns object
+func (s *DroppedColumnsStore) createState() state {
+	return &DroppedColumns{}
+}
+
+// Key returns encoded key of DroppedColumns state store
+func (s *DroppedColumnsStore) key() string {
+	// nolint:errcheck
+	bs, _ := json.Marshal(s.targetTable)
+	return adapter.DMDroppedColumnsKeyAdapter.Encode(string(bs))
+}
+
+// HasDroppedColumn returns whether the column is dropped before
+func (s *DroppedColumnsStore) HasDroppedColumn(ctx context.Context, col string, sourceTable SourceTable) bool {
+	state, err := s.Get(ctx)
+	if err != nil {
+		return false
+	}
+	droppedColumns := state.(*DroppedColumns)
+	if tbs, ok := droppedColumns.Cols[col]; ok {
+		if _, ok := tbs[sourceTable]; ok {
+			return true
+		}
+	}
+	return false
+}
+
+// AddDroppedColumns adds dropped columns to the state
+func (s *DroppedColumnsStore) AddDroppedColumns(ctx context.Context, cols []string, sourceTable SourceTable) error {
+	s.mu.Lock()
```

Comment:
can implement something like `(f *UnitStateStore) ReadModifyWrite` to avoid the repeated get/release lock of below functions

Reply:
Some behaviors are not easy to combine. For example, if the state is not found, AddDroppedColumn creates a new one, DeleteDropColumn will do noting.